apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.CommonTesting/SpannerTestDatabasePostgreSql.cs

```diff
@@ -0,0 +1,197 @@
+                    }
+                    return s_instance;
+                }
+            }
+            return s_instance;
+        }
+
+        private static readonly string s_generatedDatabaseName = IdGenerator.FromDateTime(prefix: "testdb_pg_", pattern: "yyyyMMdd't'HHmmss");
+
+        public string SpannerHost { get; } = GetEnvironmentVariableOrDefault("TEST_SPANNER_HOST", null);
+        public string SpannerPort { get; } = GetEnvironmentVariableOrDefault("TEST_SPANNER_PORT", null);
+        public string SpannerInstance { get; } = GetEnvironmentVariableOrDefault("TEST_SPANNER_INSTANCE", "spannerintegration");
+        public string SpannerDatabase { get; } = GetEnvironmentVariableOrDefault("TEST_SPANNER_POSTGRESQL_DATABASE", s_generatedDatabaseName);
+
+        // This is the simplest way of checking whether the environment variable was specified or not.
+        // It's a little ugly, but simpler than the alternatives.
+
+        /// <summary>
+        /// Returns true if the database was created just for this test, or false if the database was an existing one
+        /// specified through an environment variable.
+        /// </summary>
+        public bool Fresh => SpannerDatabase == s_generatedDatabaseName;
+
+        // Connection string including database, generated from the above properties
+        public string ConnectionString { get; }
+        // Connection string without the database, generated from the above properties
+        public string NoDbConnectionString { get; }
+        public string ProjectId { get; }
+        public DatabaseName DatabaseName { get; }
+        internal SpannerClientCreationOptions SpannerClientCreationOptions { get; }
+
+        private SpannerTestDatabasePostgreSql(string projectId)
+        {
+            TestLogger.Install();
+
+            ProjectId = projectId;
+            var builder = new SpannerConnectionStringBuilder
+            {
+                Host = SpannerHost,
+                DataSource = $"projects/{ProjectId}/instances/{SpannerInstance}",
+                EmulatorDetection = EmulatorDetection.EmulatorOrProduction
+            };
+            if (SpannerPort != null)
+            {
+                builder.Port = int.Parse(SpannerPort);
+            }
+            NoDbConnectionString = builder.ConnectionString;
+            SpannerClientCreationOptions = new SpannerClientCreationOptions(builder);
+            var databaseBuilder = builder.WithDatabase(SpannerDatabase);
+            ConnectionString = databaseBuilder.ConnectionString;
+            DatabaseName = databaseBuilder.DatabaseName;
+
+            MaybeCreateInstanceOnEmulator(projectId);
+            if (Fresh)
+            {
+                // Skip creating PostgreSQL database in Emulator, as it would fail.
+                if (!SpannerClientCreationOptions.UsesEmulator)
```

Comment:
Rather than skipping the database creation, I think it would be better to just fail hard - otherwise we're effectively pretending that we've made sure a database exists when we haven't.

Basically, we shouldn't run any tests that require Postgres if we're trying to run on the emulator, IMO. Quite how those are skipped is up for debate, but I don't think we should just continue here as if it's all fine.